### PR TITLE
[FIX] various branch management fixes

### PR DESCRIPTION
**Pull request by dylankiss**
### [[FIX] web: show branch hierarchy in switcher when no access to parent](https://github.com/odoo/odoo/pull/135455/commits/50d47a4ff3564ea08e771110cb24670dd53f1d9c)

Currently, when a user has access to a branch of a company, but not to the company itself, the branch will not show up in the company switcher. As a result, the user is not able to switch back to that branch when logged in to another company.

To fix this, we now show the whole hierarchy of the branches you have access to, disabling the companies/branches you don't have access to.

Odoo

### [[FIX] base: allow to archive companies via stat button](https://github.com/odoo/odoo/pull/135455/commits/f98acfa0136976dff8fad445cc542712b842b851)

Once a branch company has some data associated with it, it can't be deleted anymore. Since the branches could only be opened in a dialog, it was also impossible to archive them.

In order to allow (un)archiving a branch, we added a stat button on the company form to show the branches in a list view, where people can (un)archive them.

### [[FIX] account: allow creating move in branch without parent access](https://github.com/odoo/odoo/pull/135455/commits/0680edbbd9f17100ca039e64ef7d61f6b8695aa2)

When a user only has access to a branch and not to the parent, he can not create a new customer invoice or vendor bill, since we need to check the lock date on all ancestor companies (where the user might not have access to).

This fix allows the user to create a new customer invoice or vendor bill again when only having access to a branch.

Furthermore, in the above case we are also unable to assign a partner on the invoice/bill, since we are checking properties that might belong to an ancestor company we don't have access to. We also fixed that here, so a partner can be selected.

### [[FIX] base,core: relax {child,parent}_of with related fields](https://github.com/odoo/odoo/pull/135455/commits/e5837eaf90e6401daefc878b97b21088e324ff80)

In the context of branches, you need to be able to select taxes of an ancestor company on an invoice/bill, even when you don't have access to the ancestor company.

In order to do that, we relax the restriction when searching with a `parent_of` or `child_of` on a related field, so that it also includes ids of related field records you don't have access to. This should not be a problem, since in the end the search will return records of a model restricted by its own access rules.

[task-3503204](https://www.odoo.com/web#id=3503204&cids=1&menu_id=4720&action=333&active_id=967&model=project.task&view_type=form)

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by dylankiss** [comment]
> The way the hierarchy is resolved is also potentially rather expensive, the entire thing can be resolved by getting the root of every non-root company via the `parent_path` (I don't think there's a helper for that but @rco-odoo might know better) then get all their childrens in one shot.

@xmo-odoo Are you talking about the resolution of the hierarchy in `ir_http.py` or in the QWeb template?

**Review by Julien00859 — APPROVED** [review]
for python files

**Comment by tsb-odoo** [comment]
LGTM functionally speaking.

**Comment by dylankiss** [comment]
@xmo-odoo or @rco-odoo Can one of you have a look at the following commit of this PR: https://github.com/odoo/odoo/pull/135455/commits/26634d7299729d8199c426326f1236265e62951a ?

Just want to make sure I don't forget something that might go wrong here ...

File: addons/account/models/account_move.py
Comment (william-andre) [inline addons/account/models/account_move.py:1764]: We are not accessing anything on the company here, only it's id to fetch the properties... what am I missing?

File: addons/account/models/company.py
Comment (william-andre) [inline addons/account/models/company.py:350]: Putting the `sudo` on `self.parent_id` makes the code more readable, and it means that at least we have access to `self`

File: odoo/osv/expression.py
Comment (william-andre) [inline odoo/osv/expression.py:894]: I would put the sudo on line 871, for the same reasons `active_test` is put there.
It is already sudoed everywhere anyways...

File: addons/web/models/ir_http.py
Comment (william-andre) [inline addons/web/models/ir_http.py:161]: `unallowed_ancestor_companies_sudo` is already sudo'd, so is `comp`

File: odoo/addons/base/views/res_company_views.xml
Comment (william-andre) [inline odoo/addons/base/views/res_company_views.xml:89]: Do we really need this if we are already in a tree view? One can already use the generic "(Un)Archive" actions

File: addons/account/models/account_move.py
Comment (dylankiss) [reply addons/account/models/account_move.py:1764]: Good question. On any of these partner accesses I got the following error message, which was solved by using sudo for them.

Odoo_-_New

After further checking, I found out that the cause of the problem was the line at the top of this function doing
```py
self = self.with_company(self.journal_id.company_id)
```
When I remove that line, it seems to work fine. But I guess it is there for a reason ...

File: addons/web/models/ir_http.py
Comment (dylankiss) [reply addons/web/models/ir_http.py:161]: Totally right ;)

File: odoo/addons/base/views/res_company_views.xml
Comment (dylankiss) [reply odoo/addons/base/views/res_company_views.xml:89]: It was requested by Benj, so @tsb-odoo your call ;)

File: addons/account/models/company.py
Comment (dylankiss) [reply addons/account/models/company.py:350]: It looks cleaner indeed. It's less obvious that way though that we might be working with a sudo() record, but since we are calling the same method, I guess it is clear enough.

File: odoo/osv/expression.py
Comment (dylankiss) [reply odoo/osv/expression.py:894]: Hmm, there are still a lot of accesses where sudo is not needed, so it seems a bit overkill to have everything sudo'ed. (5 without sudo vs. 3 with sudo)

File: addons/web/models/ir_http.py
Comment (william-andre) [inline addons/web/models/ir_http.py:159]: outdated diff

File: addons/account/models/account_move.py
Comment (william-andre) [reply addons/account/models/account_move.py:1764]: > When I remove that line, it seems to work fine. But I guess it is there for a reason ...

Indeed, it is for when you create an invoice for a company that is not selected as the main one in the company selector.

One solution could be to use `self.journal_id.company_id._accessible_branch()`
Still need to work out the way company properties work with the branches :sweat_smile:

File: odoo/osv/expression.py
Comment (william-andre) [reply odoo/osv/expression.py:894]: Mmmh, I'll let the ORM/security team review this anyway, but I don't see any other call to a function that can be impacted by the environment where `sudo` is not used.
The other uses are only static properties of the class where the env is not even accessible

Comment (dylankiss) [reply odoo/osv/expression.py]: Fair enough. I moved the `sudo` to the top as suggested.

Comment (ryv-odoo) [reply odoo/osv/expression.py]: @xmo-odoo , semantically I am not against at this change. It is more a security issue: now the user can indirectly (via many2one) read `parent_path` of forbidden model/record (same as for `display_name`). IMO, it's not a sensitive information, what do you think ?

If the only reason of this change is for `check_company_domain_parent_of`, maybe we can change it instead ?

File: addons/web/models/ir_http.py
Comment (dylankiss) [reply addons/web/models/ir_http.py:153]: @xmo-odoo It can, but I found this way to be both more verbose and require the least amount of changes in other code using the `allowed_companies`.

File: odoo/osv/expression.py
Comment (dylankiss) [reply odoo/osv/expression.py]: > If the only reason of this change is for `check_company_domain_parent_of`, maybe we can change it instead ?

@ryv-odoo Hmm, that's actually an interesting idea indeed. Looking into that now.

Comment (dylankiss) [reply odoo/osv/expression.py]: @ryv-odoo @xmo-odoo Changing only the `check_company_domain_parent_of ` would work, although I think that the change in `expression.py` might be correct when looking at the adapted test in `base`: `test_expression.py`.

There we can't access the middle `res.partner` in the parent-chain, but when searching `res.partner.bank` records where we should have access to all three, I have the impression we should still allow searching the bank of the middle partner. What do you think?

Comment (dylankiss) [reply odoo/osv/expression.py]: @ryv-odoo Can you check my previous comment to see whether we leave it like this or I change it to only modify `check_company_domain_parent_of `? Thanks.

File: addons/web/models/ir_http.py
Comment (rco-odoo) [inline addons/web/models/ir_http.py:139]:
```suggestion
            disallowed_ancestor_companies_sudo = user.company_ids.sudo().parent_ids - user.company_ids
            all_companies_in_hierarchy_sudo = disallowed_ancestor_companies_sudo + user.company_ids
```

File: addons/web/tests/test_session_info.py
Comment (rco-odoo) [inline addons/web/tests/test_session_info.py:19]: Please use recordsets, as this enables the whole recordset API
```suggestion
        cls.allowed_companies = cls.company_a + cls.company_b_branch + cls.company_c
        cls.disallowed_ancestor_companies = cls.company_b
```

File: addons/web/tests/test_session_info.py
Comment (rco-odoo) [inline addons/web/tests/test_session_info.py:30]: Having a recordset enables such simplifications:
```suggestion
            'company_ids': [Command.set(cls.allowed_companies.ids)],
```

File: odoo/osv/expression.py
Comment (rco-odoo) [inline odoo/osv/expression.py:871]: Suggestion: rename this variable as `left_model_sudo`, to make it crystal clear that it's sudoed everywhere.

File: odoo/addons/base/models/res_company.py
Comment (rco-odoo) [inline odoo/addons/base/models/res_company.py:398]: The purpose of this action is to show children companies.  The kind of view is a detail which might be changed later.  So I suggest a more agnostic name like `action_company_branches`.

File: odoo/addons/base/models/res_company.py
Comment (rco-odoo) [inline odoo/addons/base/models/res_company.py:406]: Wouldn't it be clearer to have `active_test=False` in the action's context (this one) instead of the context of the button?  If you ever do so, I suggest to rename the method `action_all_company_branches`.

File: odoo/addons/base/views/res_company_views.xml
Comment (rco-odoo) [inline odoo/addons/base/views/res_company_views.xml:15]: Note for the forward-port, on master this attribute will become
```suggestion
                                invisible="not all_child_ids"
```

File: odoo/addons/base/views/res_company_views.xml
Comment (rco-odoo) [inline odoo/addons/base/views/res_company_views.xml:88]: Note for the forward-port: `invisible`'s value must be Python code, so on master this will become
```suggestion
                    <attribute name="invisible">False</attribute>
```

**Review by rco-odoo — COMMENTED** [review]
Looks pretty much fine to me.

File: odoo/addons/base/views/res_company_views.xml
Comment (william-andre) [reply odoo/addons/base/views/res_company_views.xml:88]: Isn't `0` a valid python expression?

File: odoo/addons/base/views/res_company_views.xml
Comment (william-andre) [reply odoo/addons/base/views/res_company_views.xml:89]: @tsb-odoo up?

File: odoo/osv/expression.py
Comment (ryv-odoo) [reply odoo/osv/expression.py]: Agree, it seems logic to me. But the question is more for the security team: is it ok to indirectly leak parent_path of inaccessible record via one Many2one

File: odoo/addons/base/models/res_company.py
Comment (dylankiss) [reply odoo/addons/base/models/res_company.py]: Perfectly fine. I moved it to the action and updated the name.

File: odoo/addons/base/views/res_company_views.xml
Comment (dylankiss) [reply odoo/addons/base/views/res_company_views.xml]: To be consistent, I changed both this one and the `invisible="1"` higher up to their boolean counterparts.

File: odoo/osv/expression.py
Comment (dylankiss) [reply odoo/osv/expression.py]: Good idea, agreed. Changed.

File: odoo/addons/base/views/res_company_views.xml
Comment (tsb-odoo) [reply odoo/addons/base/views/res_company_views.xml]: If I remember correctly, when I first tested, the archived branches were excluded from the view. So, the user couldn't access the archived branch(es) to unarchive. Re-testing now, given the inactive branches are not filtered out from the list view, the toggle button is no more important.

Comment (dylankiss) [reply odoo/addons/base/views/res_company_views.xml]: @william-andre @tsb-odoo Alright, I removed the view with the toggle buttons.

**Comment by dylankiss** [comment]
@odoo/rd-runbot I added a variant `all_child_ids` on an existing One2many field `child_ids`, just altering the context to include inactive records, to use for conditions in a view.
```py
    child_ids = fields.One2many('res.company', 'parent_id', string='Branches')
    all_child_ids = fields.One2many('res.company', 'parent_id', context={'active_test': False})
```

**Comment by dylankiss** [comment]
@william-andre Not sure if there is still something to be addressed here?

File: odoo/osv/expression.py
Comment (william-andre) [inline odoo/osv/expression.py:882]: Doesn't really make sense to use `sudo` for `_name`, `_parent_name`, `_fields` or `_parent_store`, this is just generating noise in the diff

**Review by william-andre — APPROVED** [review]
Good for me! (except for that little detail)

I'll leave a few days for others if they still have stuff to say (about the session info?)

File: odoo/addons/base/views/res_company_views.xml
Comment (william-andre) [inline odoo/addons/base/views/res_company_views.xml:74]: And this doesn't seem to be in the right commit

File: odoo/osv/expression.py
Comment (dylankiss) [reply odoo/osv/expression.py]: True. I updated that.

File: odoo/addons/base/views/res_company_views.xml
Comment (dylankiss) [reply odoo/addons/base/views/res_company_views.xml:74]: Oops, should be corrected now

**Comment by d-fence** [comment]
Upgrade exception [#255](https://runbot.odoo.com/web/#id=255&view_type=form&model=runbot.upgrade.exception) added.

Waiting the forward-port of this pr, the exception will be applied on all builds, and create an inconsistent state for migrations.
**Please forward port this asap up to master without change.** Exception should be forward-ported before the end of the week.

If you need to apply any change before it reaches master, please notify runbot team.

Details:
```
field:res.company.all_child_ids
```
cc @KangOl @nseinlet

**Comment by william-andre** [comment]
@robodoo r+

**Comment by william-andre** [comment]
@robodoo rebase-ff

**Pull request closed without merging**
